### PR TITLE
Add DispenserUseEvent

### PR DESCRIPTION
**Pull request by Zailer43**

### CODE REVIEW

File: src/main/java/xyz/nucleoid/stimuli/event/block/DispenserUseEvent.java [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+ *
+ * <p>Upon return:
+ * <ul>
+ * <li>{@link ActionResult#SUCCESS} cancels further processing and allows the use.
+ * <li>{@link ActionResult#FAIL} cancels further processing and cancels the use.
+ * <li>{@link ActionResult#PASS} moves on to the next listener.</ul>
+ * <p>
+ * If all listeners return {@link ActionResult#PASS}, the use succeeds and proceeds with normal logic.
+ */
+public interface DispenserUseEvent {
```
Comment (haykam821) [inline src/main/java/xyz/nucleoid/stimuli/event/block/DispenserUseEvent.java:22]: Could this event have a more descriptive name? I was thinking about 'dispense' or 'fire' but 'activate' as seen in your Javadoc summary for the class seems to work well.

File: src/main/java/xyz/nucleoid/stimuli/event/block/DispenserUseEvent.java [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+                    return result;
+                }
+            }
+        } catch (Throwable t) {
+            ctx.handleException(t);
+        }
+        return ActionResult.PASS;
+    });
+
+    ActionResult onUse(ServerWorld world, BlockPos pos, BlockPointerImpl blockPointerImpl, DispenserBlockEntity dispenserBlockEntity, int slot, ItemStack itemStack);
```
Comment (haykam821) [inline src/main/java/xyz/nucleoid/stimuli/event/block/DispenserUseEvent.java:37]: The `BlockPointer` class could be used in place of the `BlockPointerImpl` class here.

Comment (haykam821) [inline src/main/java/xyz/nucleoid/stimuli/event/block/DispenserUseEvent.java:37]: Could the parameter names be cleaned up? They should be cleaned up everywhere but here is most important as IDEs will copy these parameter names for methods implementing this interface.

File: src/main/java/xyz/nucleoid/stimuli/mixin/block/DispenserBlockMixin.java [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+public class DispenserBlockMixin {
+
+    @Inject(
+            method = "dispense",
+            at = @At(
+                    value = "INVOKE",
+                    target = "Lnet/minecraft/block/DispenserBlock;getBehaviorForItem(Lnet/minecraft/item/ItemStack;)Lnet/minecraft/block/dispenser/DispenserBehavior;",
+                    shift = At.Shift.BEFORE
+            ),
+            locals = LocalCapture.CAPTURE_FAILSOFT,
```
Comment (haykam821) [inline src/main/java/xyz/nucleoid/stimuli/mixin/block/DispenserBlockMixin.java:28]: Mixins for events should fail early as events should never be skipped.

```suggestion
            locals = LocalCapture.CAPTURE_FAILHARD,
```

**Review by haykam821 — CHANGES_REQUESTED** [review]

**Review by haykam821 — CHANGES_REQUESTED** [review]

File: src/main/java/xyz/nucleoid/stimuli/event/block/DispenserUseEvent.java (hunk 1 above)
Comment (Zailer43) [reply src/main/java/xyz/nucleoid/stimuli/event/block/DispenserUseEvent.java:22]: it's ok if
  DispenserUseEvent -> DispenserActivateEvent
  onUse -> onActivate
?

File: src/main/java/xyz/nucleoid/stimuli/event/block/DispenserUseEvent.java (hunk 2 above)
Comment (Zailer43) [reply src/main/java/xyz/nucleoid/stimuli/event/block/DispenserUseEvent.java:37]: I have changed:
BlockPointerImpl blockPointerImpl -> BlockPointer blockPointer
itemStack -> stackToDispense
any other name or type that you think should change?

**Pull request merged**